src/gbinder_fmq.c

```diff
@@ -0,0 +1,794 @@
+        gsize read_offset = read_ptr % size;
+        /* The number of bytes that can be read contiguously from
+         * read_offset without wrapping around the ring buffer */
+        gsize available_to_read_contiguous = size - read_offset;
+        available = available_to_read_contiguous < available_to_read ?
+            available_to_read_contiguous : available_to_read;
+    } else {
+        available = available_to_read;
+    }
+
+    return available;
+}
+
+gsize
+gbinder_fmq_available_to_write_bytes(
+    GBinderFmq* self,
+    gboolean contiguous)
+{
+    GBinderFmqPriv* priv = gbinder_fmq_cast(self);
+    gsize available;
+
+    gsize available_to_write =
+        gbinder_fmq_get_grantor_descriptor(self, DATA_PTR_POS)->extent -
+        gbinder_fmq_available_to_read_bytes(self, FALSE);
+
+    if (contiguous) {
+        guint64 write_ptr = atomic_load_explicit(priv->write_ptr,
+            memory_order_relaxed);
+        gsize size =
+            gbinder_fmq_get_grantor_descriptor(self, DATA_PTR_POS)->extent;
+        gsize write_offset = write_ptr % size;
+        /* The number of bytes that can be written contiguously starting from
+         * write_offset without wrapping around the ring buffer */
+        gsize available_to_write_contiguous = size - write_offset;
+        available = available_to_write_contiguous < available_to_write ?
+            available_to_write_contiguous : available_to_write;
+    } else {
+        available = available_to_write;
+    }
+
+    return available;
+}
+
+GBinderFmqGrantorDescriptor*
+gbinder_fmq_create_grantors(
+    gsize queue_size_bytes,
+    gsize num_fds,
+    gboolean configure_event_flag)
+{
+    gsize num_grantors = configure_event_flag ? EVENT_FLAG_PTR_POS + 1
+                                              : DATA_PTR_POS + 1;
+
+    GBinderFmqGrantorDescriptor* grantors =
+        g_new0(GBinderFmqGrantorDescriptor, num_grantors);
+
+    gsize mem_sizes[] = {
+        sizeof(guint64),  /* read pointer counter */
+        sizeof(guint64),  /* write pointer counter */
+        queue_size_bytes, /* data buffer */
+        sizeof(guint32)   /* event flag pointer */
+    };
+    gsize grantor_pos, offset;
+
+    for (grantor_pos = 0, offset = 0; grantor_pos < num_grantors;
+            grantor_pos++) {
+        GBinderFmqGrantorDescriptor *grantor = &grantors[grantor_pos];
+        guint32 grantor_fd_index;
+        gsize grantor_offset;
+        if (grantor_pos == DATA_PTR_POS && num_fds == 2) {
+            grantor_fd_index = 1;
+            grantor_offset = 0;
+        } else {
+            grantor_fd_index = 0;
+            grantor_offset = offset;
+            offset += mem_sizes[grantor_pos];
+        }
+        grantor->fd_index = grantor_fd_index;
+        grantor->offset = (guint32)(G_ALIGN8(grantor_offset));
+        grantor->extent = mem_sizes[grantor_pos];
+    }
+    return grantors;
+}
+
+void*
+gbinder_fmq_map_grantor_descriptor(
```

Comment:
This one looks internal too, should be static

Reply:
Made all internal functions static.